lib/proxy/http_download.rb

```diff
@@ -3,35 +3,55 @@
 module Proxy
   class HttpDownload < Proxy::Util::CommandTask
     include Util
-    DEFAULT_READ_TIMEOUT = 60
     DEFAULT_CONNECT_TIMEOUT = 10
-    DEFAULT_DNS_TIMEOUT = 10
 
     def initialize(src, dst, read_timeout = nil, connect_timeout = nil, dns_timeout = nil, verify_server_cert = false)
       @dst = dst
-      wget = which("wget")
-      read_timeout ||= DEFAULT_READ_TIMEOUT
-      dns_timeout ||= DEFAULT_CONNECT_TIMEOUT
       connect_timeout ||= DEFAULT_DNS_TIMEOUT
+      args = [which('curl')]
 
-      args = ["--connect-timeout=#{connect_timeout}",
-              "--dns-timeout=#{dns_timeout}",
-              "--read-timeout=#{read_timeout}",
-              "--tries=3",
-              "--timestamping", # turn on timestamping to prevent redownloads
-              "--no-if-modified-since", # but use HTTP HEAD instead If-Modified-Since
-              "-nv", src.to_s, "-O", dst.to_s]
-      args.unshift("--no-check-certificate") unless verify_server_cert
-      args.unshift(wget)
-      super(args)
+      # no cert verification if set
+      args << "--insecure" unless verify_server_cert
+      # print nothing
+      args << "--silent"
+      # except errors
+      args << "--show-error"
+      # timeout (others were supported by wget but not by curl)
+      args << ["--connect-timeout", connect_timeout.to_s]
+      # try several times
+      args << ["--retry", "3"]
+      # with a short delay
+      args << ["--retry-delay", "10"]
+      # but exit after one hour
+      args << ["--max-time", "3600"]
+      # keep last changed file attribute
+      args << "--remote-time"
+      # only download newer files
+      args << ["--time-cond", dst.to_s]
+      # print stats in the end
+      args << [
+        "--write-out",
+        '"Task done, result: %{http_code}, size downloaded: %{size_download}b, speed: %{speed_download}b/s, time: %{time_total}ms"'
+      ]
+      # output file
+      args << ["-o", dst.to_s]
+      # follow redirects
+      args << "--location"
+      # and the url to download
+      args << src.to_s
+
+      super(args.flatten)
     end
 
     def start
-      lock = Proxy::FileLock.try_locking(@dst)
+      lock = Proxy::FileLock.try_locking("#{@dst}.lock")
```

Comment:
This made me a ton of headache this morning when I was debugging it, it confused curl so it would not download anything as the lock created was always newer than the source file. So now it creates an explicit lock file and it is deleted after download is finished.

Reply:
I'm not strictly opposed, but I wonder if we can make it a hidden file. Also wondering if it should use the target directory or rather some temporary other place.

Reply:
Sure no problem.